// besdk/pin.js
const EventEmitter = require("events")

let __flyweight = {}

class Pin extends EventEmitter {
    constructor(gpio) {
        if(__flyweight[gpio])
            return __flyweight[gpio]

        super()
        this.gpio = gpio
        this.mode = ""
        __flyweight[gpio] = this
    }

    /**
     * @param {function} callback 
     * @param {string|object|undefined} options="falling"
     */
    watch(callback, options) {
        if(!options)
            options = "falling"
        if(typeof options=="string") {
            options = {
                mode: options ,
                debounce: 10 ,
                once: false ,
            }
        }
        if(options.debounce>0) {
            let oricb = callback
            let timer = -1
            callback = (val, pin)=>{
                if(timer>-1) 
                    return
                timer = setTimeout(()=>{
                    timer = -1
                    if(digitalRead(pin.gpio) == val) {
                        oricb(val, pin)
                    }
                }, options.debounce)
            }
            callback.__origin = oricb
        }
        if(!!options.once)
            this.once(options.mode, callback)
        else 
            this.on(options.mode,callback)
        this.resetISR()
        return this
    }
    unwatch(callback, mode) {
        if(!mode) {
            this.off("rising", callback, true)
            this.off("falling", callback, true)
            this.off("both", callback, true)
        }
        else {
            this.off(mode, callback, true)
        }
        this.resetISR()
        return this
    }
    clear() {
        this.off("rising")
        this.off("falling")
        this.off("both")
        this.resetISR()
        return this
    }

    resetISR() {
        let mode = 0 ;
        if(this._handles["both"]&&this._handles["both"].length) {
            mode = 3 // both
        }
        if(this._handles["rising"]&&this._handles["rising"].length) {
            mode|= 1 // rising
        }
        if(this._handles["falling"]&&this._handles["falling"].length) {
            mode|= 2 // falling
        }
        setPinISR(this.gpio, mode)
    }

    setMode(mode) {
        pinMode(this.gpio, mode)
        this.mode = mode
        return this
    }

    write(val) {
        return digitalWrite(this.gpio, val)
    }

    read() {
        return digitalRead(this.gpio)
    }

    pwm (value, freq, resolution) {
        if(value<0 || value>1){
            throw new Error("Invalid value (0-1)")
        }
        if(freq!=undefined && resolution!=undefined) {
            pwmConfig( this.gpio, freq, resolution, Math.round(value*(2**resolution)) )
        }
        else {
            pwmWrite( this.gpio, value )
        }
    }

    pwmStop() {
        pwmStop(this.gpio)
    }
}

let _pinsISRCallbacks = []
setPinISRCallback(function(gpio){
    let val = digitalRead(gpio)
    for(let cb  of _pinsISRCallbacks) {
        cb(gpio, val? "rising": "falling", val)
    }
    if(__flyweight[gpio]) {
        __flyweight[gpio].emit(val? "rising": "falling", val, __flyweight[gpio])
        __flyweight[gpio].emit("both", val, __flyweight[gpio])
    }
})

global.watchPins = function(callback) {
    _pinsISRCallbacks.push(callback)
}

global.pin = function(gpio) {
    return new Pin(gpio)
}

// pinMode 会清除 isr 类型设置
// 可能和 pinMode() 的 c 实现有关
let _pinMode = global.pinMode
global.pinMode = function(gpio, mode) {
    _pinMode(gpio, mode)
    pin(gpio).resetISR()
}

module.exports = Pin
