// lv/index.js
require("./anchor9.js")

beapi.lvgl.Color = require("./color.js")

beapi.lvgl.Btn.prototype.setText = function setText(text) {
    if(!this._label) {
        this._label = new beapi.lvgl.Label(this)
    }
    this._label.setText(text)
}
beapi.lvgl.Obj.prototype.getFlexGrow = function getFlexGrow() {
    return this.getStyleFlexGrow(0)
}

class Row extends beapi.lvgl.Obj {
    constructor(parent) {
        super(parent)
        this.removeStyleAll()
        this.setFlexFlow("row")
    }
}
beapi.lvgl.Row = Row
class Column extends beapi.lvgl.Obj {
    constructor(parent) {
        super(parent)
        this.removeStyleAll()
        this.setFlexFlow("column")
    }
}
beapi.lvgl.Column = Column

const loader = require("./loader.js")
exports.fromJson = loader.fromJson
exports.toJson = loader.toJson

const clazzlst = [
"Obj",
"Label",
"Arc",
"Bar",
"Btn",
"BtnMatrix",
"Canvas",
"Checkbox",
"Dropdown",
"Img",
"Line",
"Roller",
"Slider",
"Switch",
"Table",
"TextArea",
"Row",
"Column",
]
for(let clzname of clazzlst) {
    exports[clzname] = beapi.lvgl[clzname]
}

beapi.lvgl.state = {
    'default': 0x0000 ,
    'checked': 0x0001 ,
    'focused': 0x0002 ,
    'focus_key': 0x0004 ,
    'edited': 0x0008 ,
    'hovered': 0x0010 ,
    'pressed': 0x0020 ,
    'scrolled': 0x0040 ,
    'disabled': 0x0080 ,
    'user_1': 0x1000 ,
    'user_2': 0x2000 ,
    'user_3': 0x4000 ,
    'user_4': 0x8000 ,
    'any': 0xFFFF ,
}
beapi.lvgl.part = {
    'main': 0x000000 ,
    'scrollbar': 0x010000 ,
    'indicator': 0x020000 ,
    'knob': 0x030000 ,
    'selected': 0x040000 ,
    'items': 0x050000 ,
    'ticks': 0x060000 ,
    'cursor': 0x070000 ,
    'custom_first': 0x080000 ,
    'any': 0x0F0000 ,
}
