// lv/loader.js
const MapPropFuncs = {
    align: 'setAlign' ,
    flex: 'setFlexFlow' ,
    flexAlign: 'setFlexAlign' ,
    text: 'setText' ,
    grow: 'setFlexGrow' ,
}

exports.fromJson = function fromJson(json, parent, refs){

    if(json instanceof Array) {
        for(let childJson of json) {
            fromJson(childJson, obj, refs)
        }
    }
    else {

        let clzname = json["class"] || json.clazz
        if(!clzname) {
            throw new Error("missing class name")
        }
        if(!beapi.lvgl[clzname]){
            throw new Error("unknow class name: "+clzname)
        }
        let obj = new beapi.lvgl[clzname](parent)

        if(!refs)
            refs={}
        if(json.name) {
            refs[json.name] = obj
        }

        for(let propName in MapPropFuncs) {
            if(typeof json[propName]=="undefined")
                continue
            let methodName = MapPropFuncs[propName]
            if(typeof obj[methodName]=='function') {
                console.log("invoke", methodName, json[propName])
                obj[methodName]( json[propName] )
            }
        }

        if(json.children) {
            for(let childJson of json.children) {
                fromJson(childJson, obj, refs)
            }
        }
    }
    
    return refs
}
exports.toJson = function fromJson(){
    throw new Error("not implemented")
}
